lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +434,1140 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+        Result.insert({ CI, FunctionEdgeTypeValue::IndirectTailCall });
+      else
+        Result.insert({ CI, FunctionEdgeTypeValue::LongJmp });
+    }
+  }
+
+  bool FoundRet, FoundBrokenRet;
+  FoundRet = FoundBrokenRet = false;
+  for (const auto &E : Result) {
+    if (E.second == FunctionEdgeTypeValue::Return)
+      FoundRet = true;
+    if (E.second == FunctionEdgeTypeValue::BrokenReturn)
+      FoundBrokenRet = true;
+  }
+
+  FunctionTypeValue Type;
+  if (FoundRet) {
+    Type = FunctionTypeValue::Regular;
+  } else if (FoundBrokenRet) {
+    Type = FunctionTypeValue::Fake;
+  } else {
+#ifdef NORETURN_IS_FIXED
+    Type = FunctionTypeValue::NoReturn;
+#else
+    Type = FunctionTypeValue::Regular;
+#endif
+  }
+
+  // Finally retrieve the clobbered registers and return them.
+  std::erase_if(ClobberedRegs, [&](const auto &E) {
+    return CalleeSavedRegs.find(E) != CalleeSavedRegs.end();
+  });
+
+  // Finalize CFG for the model.
+  for (const auto &[CI, EdgeValue] : Result) {
+    auto PC = MetaAddress::fromConstant(CI->getArgOperand(2));
+    auto &Block = CFG.at(PC);
+    Block.Successors.insert(makeEdge(MetaAddress::invalid(), EdgeValue));
+  }
+
+  // We say that a register is callee-saved when, besides being preserved by the
+  // callee, there is at least a write onto this register.
+  std::set<GlobalVariable *> ActualCalleeSavedRegs;
+  std::set_intersection(CalleeSavedRegs.begin(),
+                        CalleeSavedRegs.end(),
+                        WrittenRegisters.begin(),
+                        WrittenRegisters.end(),
+                        std::inserter(ActualCalleeSavedRegs,
+                                      ActualCalleeSavedRegs.begin()));
+
+  // Union between effective callee-saved registers and SP.
+  std::set<GlobalVariable *> CalleeSavedAndStackPointerRegs = {
+    ActualCalleeSavedRegs
+  };
+  CalleeSavedAndStackPointerRegs.insert(GCBI->spReg());
+
+  // Refine ABI analyses by suppressing callee-saved and stack pointer registers
+  // from the final results.
+  for (const auto &Reg : CalleeSavedAndStackPointerRegs) {
+    if (ABIResults.Arguments.count(Reg) != 0)
+      ABIResults.Arguments[Reg] = model::RegisterState::Values::No;
+  }
+
+  for (const auto &[K, _] : ABIResults.CallSites) {
+    for (const auto &Reg : CalleeSavedAndStackPointerRegs) {
+      if (ABIResults.CallSites[K].Arguments.count(Reg) != 0)
+        ABIResults.CallSites[K]
+          .Arguments[Reg] = model::RegisterState::Values::No;
+
+      if (ABIResults.CallSites[K].ReturnValues.count(Reg) != 0)
+        ABIResults.CallSites[K]
+          .ReturnValues[Reg] = model::RegisterState::Values::No;
+    }
+  }
+
+  for (const auto &E : Result) {
+    if (E.second != FunctionEdgeTypeValue::Return
+        && E.second != FunctionEdgeTypeValue::IndirectTailCall) {
+      auto PC = MetaAddress::fromConstant(E.first->getOperand(2));
+
+      auto It = ABIResults.ReturnValues.find(PC);
```

Comment:
I'm confused by `ReturnValues`, more like `ReturnInstructionAddresses`?

Reply:
It's a map (`std::map<MetaAddress, RegisterStateMap>`) from ret instruction address to the map including the CSVs ABI registers return values for that ret instruction. `ReturnValuesForRetInstruction` makes more sense?

Reply:
A "return value" is 3  in `return 3`. It doesn't sound appropriate.
What about `ReturnInstructionsState`?

Reply:
Should we change `FinalReturnValues` into `FinalReturnInstructionsState` too? `FinalReturnValues` is simply a  `RegisterStateMap` with all the combined return value registers.

Reply:
Mmmh, maybe `ReturnValues` is not so bad after all... If we have `Arguments` instead of `ArgumentsLocations`  we can also have `ReturnValues` instead of `ReturnValuesLocations`, which would be the appropriate thing to do but it's a bit verbose.

Up to you, you can leave it as is or switch everything to `ArgumentsLocations`/`ArgumentsRegisters` and `ReturnValuesLocation`/`ReturnValuesRegisters`.

Reply:
Moved to `ArgumentsRegisters` / `ReturnValuesRegisters`, maybe a bit verbose, but it must be clear IMHO we are talking about registers.

Reply:
Remember to push things that you state you did, I can't see it.

Reply:
Yeah sorry, I was really about to.